bgpd/bgp_mplsvpn.c

```diff
@@ -233,6 +236,762 @@ int bgp_nlri_parse_vpn(struct peer *peer, struct attr *attr,
+		if (debug)
+			zlog_debug("%s: wrong type %d", __func__,
+				   info_vpn->type);
+		return;
+	}
+	if (info_vpn->sub_type != BGP_ROUTE_NORMAL
+	    && info_vpn->sub_type != BGP_ROUTE_STATIC) {
+
+		if (debug)
+			zlog_debug("%s: wrong sub_type %d", __func__,
+				   info_vpn->sub_type);
+		return;
+	}
+
+	if (debug)
+		zlog_debug("%s: start (info_vpn=%p)", __func__, info_vpn);
+
+
+	/* Loop over VRFs */
+	for (ALL_LIST_ELEMENTS(bm->bgp, mnode, mnnode, bgp)) {
+		if (!vpn_leak_from_vpn_active(bgp, afi, &debugmsg, &red)) {
+			if (debug)
+				zlog_debug("%s: skipping: %s", __func__,
+					   debugmsg);
+			continue;
+		}
+
+		/* Check for intersection of route targets */
+		if (!ecom_intersect(bgp->vpn_policy[afi]
+					    .rtlist[BGP_VPN_POLICY_DIR_FROMVPN],
+				    info_vpn->attr->ecommunity)) {
+
+			continue;
+		}
+
+		if (debug)
+			zlog_debug("%s: withdrawing from vrf %s", __func__,
+				   bgp->name);
+
+		bn = bgp_afi_node_get(bgp->rib[afi][safi], afi, safi, p, NULL);
+		for (bi = (bn ? bn->info : NULL); bi; bi = bi->next) {
+			if (bi->extra
+			    && (struct bgp_info *)bi->extra->parent
+				       == info_vpn) {
+				break;
+			}
+		}
+
+		if (bi) {
+			if (debug)
+				zlog_debug("%s: deleting bi %p", __func__, bi);
+			bgp_aggregate_decrement(bgp, p, bi, afi, safi);
+			bgp_info_delete(bn, bi);
+			bgp_process(bgp, bn, afi, safi);
+		}
+		bgp_unlock_node(bn);
+	}
+}
+
+void vpn_leak_to_vrf_withdraw_all(struct bgp *bgp_vrf, /* to */
+				  afi_t afi)
+{
+	struct bgp_node *bn;
+	struct bgp_info *bi;
+	safi_t safi = SAFI_UNICAST;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_TO_VRF);
+
+	if (debug)
+		zlog_debug("%s: entry", __func__);
+	/*
+	 * Walk vrf table, delete bi with type/subtype
+	 * ZEBRA_ROUTE_BGP_VPN/BGP_ROUTE_REDISTRIBUTE
+	 */
+	for (bn = bgp_table_top(bgp_vrf->rib[afi][safi]); bn;
+	     bn = bgp_route_next(bn)) {
+
+		for (bi = bn->info; bi; bi = bi->next) {
+			if (bi->type == ZEBRA_ROUTE_BGP_VPN
+			    && bi->sub_type == BGP_ROUTE_REDISTRIBUTE) {
+
+				/* delete route */
+				bgp_aggregate_decrement(bgp_vrf, &bn->p, bi,
+							afi, safi);
+				bgp_info_delete(bn, bi);
+				bgp_process(bgp_vrf, bn, afi, safi);
+			}
+		}
+	}
+}
+
+void vpn_leak_to_vrf_update_all(struct bgp *bgp_vrf, /* to */
+				struct bgp *bgp_vpn, /* from */
+				afi_t afi)
+{
+	struct prefix_rd prd;
+	struct bgp_node *prn;
+	safi_t safi = SAFI_MPLS_VPN;
+
+	/*
+	 * Walk vpn table
+	 */
+	for (prn = bgp_table_top(bgp_vpn->rib[afi][safi]); prn;
+	     prn = bgp_route_next(prn)) {
+
+		struct bgp_table *table;
+		struct bgp_node *bn;
+		struct bgp_info *bi;
+
+		memset(&prd, 0, sizeof(prd));
+		prd.family = AF_UNSPEC;
+		prd.prefixlen = 64;
+		memcpy(prd.val, prn->p.u.val, 8);
+
+		/* This is the per-RD table of prefixes */
+		table = prn->info;
+
+		if (!table)
+			continue;
+
+		for (bn = bgp_table_top(table); bn; bn = bgp_route_next(bn)) {
+
+			for (bi = bn->info; bi; bi = bi->next) {
+
+				if (bi->type == ZEBRA_ROUTE_BGP_VPN)
```

Comment:
This is mentioned in my general comment too, but in addition, this won't work for VRF-route-leaking because the route exported from (one) VRF to the VPN table is of type ZEBRA_ROUTE_BGP_VPN and this check will prevent it being imported into another VRF - unless I misunderstood.

Reply:
Agreed. For now I have added a check also between the source and destination vrf, which should permit vrf1->vpn->vrf2.